analysis.go

```diff
@@ -36,222 +39,141 @@ type Info struct {
-	return checkProjectTask(func() []CheckTask {
-		return []CheckTask{CheckEventLogForErrors, CheckDeployConfigsReplicasNotZero, CheckForWaitingPods}
-	}, JSONResourceFactory, project, results)
-}
-
-// A getProjectCheckFactory generates tasks to diagnose system conditions.
-type getProjectCheckFactory func() []CheckTask
-
-// checkTasks executes all the CheckTasks returned from the supplied
-// checkFactory against the specified project. The results of the checks are
-// combined into a single JSON object and returned.
-func checkProjectTask(checkFactory getProjectCheckFactory, JSONResourceFactory DumpedJSONResourceFactory, project string, results chan<- AnalysisResult) Task {
+// CheckProjectTask returns a task that diagnoses problems in the project scope.
+func CheckProjectTask(project string, definition DefinitionLoader, results chan<- AnalysisResult) Task {
 	return func() error {
 		result := ProjectResult{Project: project}
-		checks := checkFactory()
 
-		var errors errorList
-		for _, check := range checks {
-			res, err := check(JSONResourceFactory)
-			if err != nil {
-				errors = append(errors, err)
-			}
-			result.Results = append(result.Results, res)
+		var (
+			events            types.EventList
+			deploymentConfigs types.DeploymentConfigList
+			pods              types.PodList
+		)
+
+		definition.Load("events", &events)
+		definition.Load("deploymentconfigs", &deploymentConfigs)
+		definition.Load("pods", &pods)
 
+		if err := definition.Err(); err != nil {
+			return err
 		}
 
+		result.Results = append(result.Results,
+			CheckEvents(events),
+			CheckDeploymentConfigs(deploymentConfigs),
+			CheckPods(pods),
+		)
+
 		results <- AnalysisResult{Projects: []ProjectResult{result}}
 
-		if len(errors) > 0 {
-			return errors
-		}
 		return nil
 	}
 }
 
-// DumpedJSONResourceFactory takes a string containing the path to a
-// JSON file which will be parsed and loaded into the supplied interface.
-type DumpedJSONResourceFactory func(string, interface{}) error
+// A DefinitionLoader loads definitions of OpenShift resources.
+type DefinitionLoader interface {
+	Load(kind string, v interface{})
+	Err() error
+}
 
-// getDumpedJSONResourceFactory returns a factory which will load resource from
-// a given basepath. The factory parses the file contents as JSON and loads it
-// into the provided dest interface.
-func getDumpedJSONResourceFactory(basepath string) DumpedJSONResourceFactory {
-	return func(path string, dest interface{}) error {
-		contents, err := ioutil.ReadFile(filepath.Join(basepath, path))
-		if err != nil {
-			return err
-		}
-		return json.Unmarshal(contents, dest)
+// definitionLoader loads JSON resource definitions from files.
+type definitionLoader struct {
+	basepath, project string
+	err               error
+}
+
+func (l *definitionLoader) Load(kind string, v interface{}) {
+	if l.err != nil {
+		return
+	}
+	path := filepath.Join(l.basepath, "projects", l.project, "definitions", kind+".json")
```

Comment:
One advantage of being a `definitionLoader` is that we can talk in terms of the `DefinitionLoader` interface, which could really be called `Loader` if we were to have specific loaders for other things.

What bothers me more here is that we have the knowledge about where definitions are stored in two different places: here and where they are written (by the DumpRunner).